src/gpacmp4/mp4.c

```diff
@@ -304,20 +306,178 @@ unsigned char * ccdp_find_data(unsigned char * ccdp_atom_content, unsigned int l
+				unsigned char cc_valid = (unsigned char)((cc_info & 4) >> 2);
+				unsigned char cc_type = (unsigned char)(cc_info & 3);
+
+				if (cc_info == CDP_SECTION_SVC_INFO || cc_info == CDP_SECTION_FOOTER) {
+					dbg_print(CCX_DMT_PARSE, "MP4-708: premature end of sample (0x73 or 0x74)\n");
+					break;
+				}
+
+				if ((cc_info == 0xFA || cc_info == 0xFC || cc_info == 0xFD)
+					&& (cc_data[1] & 0x7F) == 0 && (cc_data[2] & 0x7F) == 0) {
+					dbg_print(CCX_DMT_PARSE, "MP4-708: skipped (zero cc data)\n");
+					continue;
+				}
+
+				temp[0] = cc_valid;
+				temp[1] = cc_type;
+				temp[2] = cc_data[1];
+				temp[3] = cc_data[2];
+
+				if (cc_type < 2) {
+					dbg_print(CCX_DMT_PARSE, "MP4-708: atom skipped (cc_type < 2)\n");
+					continue;
+				}
+				dec_ctx->dtvcc->encoder = (void *)enc_ctx; //WARN: otherwise cea-708 will not work
+														   //TODO is it really always 4-bytes long?
+				ccx_dtvcc_process_data(dec_ctx, (unsigned char *)temp, 4);
+				cb_708++;
+			}
+		}
+		else //subtype == GF_ISOM_SUBTYPE_C608
+		{
+			if (is_ccdp) {
+				mprint("Your video file seems to be an interesting sample for us\n");
+				mprint("We haven't seen a c608 subtitle inside a ccdp atom before\n");
+				mprint("Please, report\n");
+				return -1;
+			}
+
+			int ret = 0;
+			int len = atom_length - 8;
+			data += 4;
+			char *tdata = data;
+			do {
+				// Process each pair independently so we can adjust timing
+				ret = process608((unsigned char *)tdata, len > 2 ? 2 : len, dec_ctx, dec_sub);
+				len -= ret;
+				tdata += ret;
+				cb_field1++;
+				if (dec_sub->got_output) {
+					*mp4_ret = 1;
+					encode_sub(enc_ctx, dec_sub);
+					dec_sub->got_output = 0;
+				}
+			} while (len > 0);
+		}
+	}
+	return atom_length;
+}
+
+// Rrocess tx3g type atom
+// Return the length of the atom
+// Return -1 if unrecoverable error happened or process of the atom is finished.
+// In this case, the sample will be skipped.
+// Argument `encode_last_only` is used to print the last subtitle
+static int process_tx3g(struct lib_ccx_ctx *ctx, struct encoder_ctx *enc_ctx,
+	struct lib_cc_decode *dec_ctx, struct cc_subtitle *dec_sub, int *mp4_ret,
+	char *data, size_t data_length, int encode_last_only) {
+	// tx3g data format:
+	// See https://developer.apple.com/library/content/documentation/QuickTime/QTFF/QTFFChap3/qtff3.html (Section Subtitle Sample Data)
+	// See http://www.etsi.org/deliver/etsi_ts/126200_126299/126245/14.00.00_60/ts_126245v140000p.pdf (Section 5.17)
+	// Basically it's 16-bit length + UTF-8/UTF-16 text + optional extension (ignored)
+	// TODO: support extension & style
+
+	static struct cc_subtitle last_sub;
+	static int has_previous_sub = 0;
+
+	// Always encode the previous subtitle, no matter the current one is valid or not
+	if (has_previous_sub) {
+		dec_sub->end_time = dec_ctx->timing->fts_now;
+		encode_sub(enc_ctx, dec_sub); // encode the previous subtitle
+		has_previous_sub = 0;
+	}
+	if (encode_last_only) return;
```

Comment:
@saurabhshri 
Sorry that I missed the review till now.
The purpose of this `if` block is merely to exit the function early so that it won't attempt to process more data. (since it's the last one)
And it seems that it has already been fixed.